ledger/acctupdates.go

```diff
@@ -416,6 +417,138 @@ func (au *accountUpdates) lookupKv(rnd basics.Round, key string, synchronized bo
+	return au.lookupKeysByPrefix(round, keyPrefix, maxKeyNum, true /* take lock */)
+}
+
+func (au *accountUpdates) lookupKeysByPrefix(
+	round basics.Round, keyPrefix string, maxKeyNum uint64, synchronized bool,
+) (resultKeys []string, err error) {
+
+	// define the set of key with prefix to maintain
+	keysWithValue := map[string]bool{}
+
+	needUnlock := false
+	if synchronized {
+		au.accountsMu.RLock()
+		needUnlock = true
+	}
+	defer func() {
+		if needUnlock {
+			au.accountsMu.RUnlock()
+		}
+		// preparation of result happens in deferring function
+		// prepare result only when err != nil
+		if err == nil {
+			resultKeys = make([]string, 0, len(keysWithValue))
+			for resKey := range keysWithValue {
+				resultKeys = append(resultKeys, resKey)
+			}
+		}
+	}()
+
+	// TODO: This loop and round handling is copied from other routines like
+	// lookupResource. I believe that it is overly cautious, as it always reruns
+	// the lookup if the DB round does not match the expected round. However, as
+	// long as the db round has not advanced too far (greater than `rnd`), I
+	// believe it would be valid to use. In the interest of minimizing changes,
+	// I'm not doing that now.
+
+	for {
+		currentDBRound := au.cachedDBRound
+		currentDeltaLen := len(au.deltas)
+		offset, _err := au.roundOffset(round)
+		if _err != nil {
+			err = _err
+			return
+		}
+		// reset set `keysWithValue` to be empty each iteration
+		// if db round does not match unexpected and starts over again
+		keysWithValue = map[string]bool{}
+		// keep track of deleted key in later round
+		// do not add key to `keysWithValue` set if the key exists in `sawDeletedKeys`
+		sawDeletedKeys := map[string]bool{}
+		for i := offset - 1; i > 0; i-- {
+			for keyInRound, valOp := range au.kvDeltas[i] {
+				if !strings.HasPrefix(keyInRound, keyPrefix) {
+					continue
+				}
+				if valOp == nil {
+					// if the delta deletes a keyInRound-value
+					// but in later round the keyInRound is set to a value
+					// we ignore
+					if _, ok := keysWithValue[keyInRound]; ok {
+						continue
+					}
+					// if the keyInRound is not added with value in later rounds
+					// no matter how many times it is deleted
+					// we put the keyInRound into set `sawDeletedKeys`
+					sawDeletedKeys[keyInRound] = true
+				} else {
+					// if the delta has keyInRound set to valOp
+					// but in later round the keyInRound is deleted
+					// we just ignore them
+					if _, ok := sawDeletedKeys[keyInRound]; ok {
+						continue
+					}
+					// if the keyInRound is not deleted in later rounds
+					// no matter how many times it is set
+					// we put the keyInRound into set `keysWithValue`
+					keysWithValue[keyInRound] = true
+
+					// check if the size of `keysWithValue` goes to `maxKeyNum`
+					// if so just return the list of keys
+					if maxKeyNum > 0 && len(keysWithValue) == int(maxKeyNum) {
+						return
+					}
+				}
```

Comment:
```suggestion
                                 if _, ok := results[keyInRound]; ok {
                                    continue
                                 }
				if valOp == nil {
                                     results[keyInRound] = false
				} else {
				    results[keyInRound] = true;
				    resultCount++
   				    if maxKeyNum > 0 && resultCount == maxKeyNum {
				       return
				    }
				}
```

Reply:
just modified to adapt to the proposed implementation.